parser/prism/Translator.cc

```diff
@@ -1348,12 +1348,38 @@ unique_ptr<parser::Node> Translator::patternTranslate(pm_node_t *node) {
         case PM_IN_NODE: { // An `in` pattern such as in a `case` statement, or as a standalone expression.
             auto inNode = reinterpret_cast<pm_in_node *>(node);
 
-            auto sorbetPattern = patternTranslate(inNode->pattern);
-
+            auto prismPattern = move(inNode->pattern);
             auto inlineIfSingle = true;
             auto statements = translateStatements(inNode->statements, inlineIfSingle);
 
-            return make_unique<parser::InPattern>(location, move(sorbetPattern), nullptr, move(statements));
+            if (prismPattern != nullptr &&
+                (PM_NODE_TYPE_P(prismPattern, PM_IF_NODE) || PM_NODE_TYPE_P(prismPattern, PM_UNLESS_NODE))) {
+                pm_statements_node *prismStatements = nullptr;
+                std::unique_ptr<parser::Node> sorbetGuard;
+
+                if (PM_NODE_TYPE_P(prismPattern, PM_IF_NODE)) {
+                    auto ifNode = reinterpret_cast<pm_if_node *>(prismPattern);
+                    prismStatements = reinterpret_cast<pm_statements_node *>(ifNode->statements);
+                    sorbetGuard = std::make_unique<parser::IfGuard>(location, translate(ifNode->predicate));
+                } else { // PM_UNLESS_NODE
+                    auto unlessNode = reinterpret_cast<pm_unless_node *>(prismPattern);
+                    prismStatements = reinterpret_cast<pm_statements_node *>(unlessNode->statements);
+                    sorbetGuard = std::make_unique<parser::UnlessGuard>(location, translate(unlessNode->predicate));
+                }
+
+                if (prismStatements->body.size != 1) {
+                    unreachable("In pattern-matching's `in` clause, a guard must have a single statement.");
+                }
```

Comment:
```suggestion
                ENFORCE(prismStatements->body.size == 1,
                    "In pattern-matching's `in` clause, a guard must have a single statement.");
```

Reply:
Enforcing this is good for now, but we'll eventually want to tolerate this error and report it as a syntax error.

This is the current Sorbet's parser behaviour:

```ruby
  case quux
# ^^^^ error: unexpected token "case"
  in x if 1, 2
  end
```

Interestingly, the `, 2` makes the `case` invalid, not the `in` or `if`. Removing that makes it parse correctly.

Reply:
Are you sure it's not the `1, 2`? Maybe I'm misunderstanding.

Reply:
Yes I'm sure because the `prismStatements` is referring to the statements under the `IfNode`, not the `InNode`:

```rb
case bar
in x if 1, 2
end
```

<img width="80%%" alt="Screenshot 2024-10-24 at 22 23 10" src="https://github.com/user-attachments/assets/c8692289-107e-4e85-a251-350e95c420f8">

I will rename the variable to make that clearer.